test-e2e/test_master_node_replacement.py

```diff
@@ -0,0 +1,209 @@
+
+            $IP_CMD -4 -o addr show dev {interface} | awk '{{split($4,a,"/");print a[1]}}'
+            """.format(interface=interface),
+        )
+        ip_detect_path = tmp_path / 'ip-detect'
+        ip_detect_path.write_text(data=ip_detect_contents)
+        static_config = {
+            'master_discovery': 'static',
+            'master_list': [str(master.private_ip_address)
+                            for master in original_cluster.masters],
+        }
+        dcos_config = {
+            **original_cluster.base_config,
+            **static_config,
+        }
+        original_cluster.install_dcos_from_path(
+            dcos_installer=artifact_path,
+            dcos_config=dcos_config,
+            ip_detect_path=ip_detect_path,
+        )
+        wait_for_dcos_oss(
+            cluster=original_cluster,
+            request=request,
+            log_dir=log_dir,
+        )
+        current_cluster = original_cluster
+        tmp_clusters = set([])
+
+        original_masters = original_cluster.masters
+
+        try:
+            for master_to_be_replaced in original_masters:
+                # Destroy a master and free one IP address.
+                original_cluster.destroy_node(node=master_to_be_replaced)
+
+                temporary_cluster = Cluster(
+                    cluster_backend=docker_backend,
+                    # Allocate one container with the now free IP address.
+                    masters=1,
+                    agents=0,
+                    public_agents=0,
+                )
+                tmp_clusters.add(temporary_cluster)
+
+                # Install a new master on a new container with the same IP address.
+                (new_master, ) = temporary_cluster.masters
+                new_master.install_dcos_from_path(
+                    dcos_installer=artifact_path,
+                    dcos_config=dcos_config,
+                    role=Role.MASTER,
+                    ip_detect_path=ip_detect_path,
+                )
+                # Form a new cluster with the newly create master node.
+                new_cluster = Cluster.from_nodes(
+                    masters=current_cluster.masters.union(set([new_master])),
+                    agents=current_cluster.agents,
+                    public_agents=current_cluster.public_agents,
+                )
+                # The `wait_for_dcos_oss` function waits until the new master has
+                # joined the cluster and all masters are healthy. Without the
+                # cockroachdb check, this succeeds before all cockroachdb ranges
+                # have finished replicating to the new master. That meant that the
+                # next master would be replaced too quickly, while it had data that
+                # was not present elsewhere in the cluster. This lead to
+                # irrecoverable dataloss.  This function waits until the node is
+                # "healthy". It is assumed that this performs the same diagnostics
```

Comment:
Can we be more definite here and remove the assumption? "This performs the same diagnostics..."